### PR TITLE
Fix unindent (shift-tab) on column 0

### PR DESCRIPTION
**Pull request by ghost**
Fixes unindenting (pressing shift-tab) in this situation:

bildschirmfoto 2017-12-23 um 09 27 27

In current master, nothing happens.

### CODE REVIEW

**Comment by poke1024** [comment]
In addition to the "column 0" fix, these are actually now 3 changes to the unindent, as seen in this video:

https://www.dropbox.com/s/f17aqi0qrxwxgy5/demo-unindent.mp4?dl=0

(1) unindent in spaced area left does no longer move cursor
(2) unindenting also works if the cursor is not in spaced area
(3) unindenting with a number of spaces < indent size will remove those spaces

These three are very common unindenting behaviors in code editors (e.g. in all JetBrains products).

File: scene/gui/text_edit.cpp [hunk 1]
```diff
@@ -2420,26 +2420,44 @@ void TextEdit::_gui_input(const Ref<InputEvent> &p_gui_input) {
 
 						//simple unindent
 						int cc = cursor.column;
+
+						const int len = text[cursor.line].length();
+						const String &line = text[cursor.line];
+
+						int left;
```
Comment (Paulb23) [inline scene/gui/text_edit.cpp]: Left is not always defined causing crashes and weird behaviour. Should be initialised to 0

File: scene/gui/text_edit.cpp [hunk 2]
```diff
@@ -2420,26 +2420,44 @@ void TextEdit::_gui_input(const Ref<InputEvent> &p_gui_input) {
...
+								}
 
-									if (unindent) {
-										_remove_text(cursor.line, cursor.column - indent_size, cursor.line, cursor.column);
-										cursor_set_column(cursor.column - indent_size);
-									}
+								if (n > 0) {
+									_remove_text(cursor.line, cc - n, cursor.line, cc);
+									if (cursor.column > left - n)
+										cursor_set_column(MAX(0, left - n));
```
Comment (Paulb23) [inline scene/gui/text_edit.cpp]: Can this not be `cursor.column - n` to keep caret position when unindenting?

File: scene/gui/text_edit.cpp (hunk 1 above)
Comment (poke1024) [reply scene/gui/text_edit.cpp]: Indeed, missed that completely.

File: scene/gui/text_edit.cpp (hunk 2 above)
Comment (poke1024) [reply scene/gui/text_edit.cpp]: No, but the code is confusing - I tried to work on the existing code, which always deletes to the left of the cursor and things would be simpler if we'd delete to the right, but that would have changed about everything.

Here are the two cases that can happen:

Case 1:

Take the following text (X is space, T is text, ^ is the cursor position, and `left` is the number of `X`):

```
XXXXXXXXXXTTTTTTT
    ^
```

An unindent happens; `_remove_text` will remove the area to the left of the cursor marked by R:

```
RRRRXXXXXXTTTTTTT
    ^
```

And after removing:

```
XXXXXXTTTTTTT
    ^
```

In this case we do not want to move the cursor, as it is in the white space area and should stay at the same position as before.

Case 2:

Take this text:

```
XXXXXXXXXXTTTTTTT
        ^
```

Removing:

```
XXXXRRRRXXTTTTTTT
        ^
```

```
XXXXXXTTTTTTT
      * ^
```

Now the cursor is in the text area (which begins at `left - n` at `*`) and we need to move the cursor left.

Comment (Paulb23) [reply scene/gui/text_edit.cpp]: Should've been more clear, I meant the `cursor_set_column(MAX(0, left - n));` to `cursor_set_column(MAX(0, cursor.column - n));` else the caret is always going to be moved the start of the text. :)

Comment (poke1024) [reply scene/gui/text_edit.cpp]: Ah ok, that makes sense, changed.

**Pull request merged**